File: examples/python/tbr.py
"""
Tritium Breeding Ratio (TBR) example with per-nuclide breakdown.

Compares total H3-production against per-nuclide contributions from Li6, Li7,
and Be9. The sum of per-nuclide values should match the total within statistical
uncertainty.
"""

import time

import yamc

# --- Geometry: point source inside a Li/Be sphere ---
sphere1 = yamc.Sphere(x0=0, y0=0, z0=0, radius=1.0)
sphere2 = yamc.Sphere(x0=0, y0=0, z0=0, radius=200.0,
                    boundary='vacuum')
region1 = sphere1.below
region2 = sphere1.above & sphere2.below

# --- Material: Li4SiO4 surrogate (Li6 + Li7 + Be9) ---
material1 = yamc.Material(
    composition={
        "Li6": 0.07 / 2,
        "Li7": 0.93 / 2,
        "Be9": 0.5,
    },
    density=2.0,
    temperature=294)
material1.read_nuclear_data({
    "Be9": "tests/Be9.arrow",
    "Li6": "tests/Li6.arrow",
    "Li7": "tests/Li7.arrow",
})

cell1 = yamc.Cell(name="inner_sphere", region=region1)
cell2 = yamc.Cell(name="outer_annular", region=region2,
                material=material1)
geometry = yamc.Geometry([cell1, cell2])

source = yamc.NeutronSource(
    energy=yamc.sources.fusion_neutron_spectrum(20000.0),
    position=(0, 0, 0))

# --- Tally 1: Total H3-production (no nuclide breakdown) ---
tally_total = yamc.Tally(name="tbr_total", cells=cell2, scores=['H3-production'])

# --- Tally 2: Per-nuclide H3-production ---
tally_nuclides = yamc.Tally(
    name="tbr_per_nuclide",
    cells=cell2,
    scores=['H3-production'],
    nuclides=['Li6', 'Li7', 'Be9'],
)

tallies = [tally_total, tally_nuclides]
model = yamc.Model(geometry=geometry, tallies=tallies, source=source)

t0 = time.time()
results = model.simulate_transport(total_particles=250000, seed=1)
elapsed = time.time() - t0

# --- Print results ---
print(f"\n{'='*50}")
print(f"  TBR Results  ({elapsed:.2f}s)")
print(f"{'='*50}")

# Total H3-production
total_mean = results[tally_total].mean
if hasattr(total_mean, 'flatten'):
    total_mean = total_mean.flatten()
print(f"\n  Total H3-production: {total_mean[0]:.6e}")

# Per-nuclide breakdown
nuc_result = results[tally_nuclides]
nuc_mean = nuc_result.mean
if hasattr(nuc_mean, 'flatten'):
    nuc_mean = nuc_mean.flatten()

nuc_std = nuc_result.standard_deviation
if hasattr(nuc_std, 'flatten'):
    nuc_std = nuc_std.flatten()

nuclide_names = ['Li6', 'Li7', 'Be9']
print("\n  Per-nuclide H3-production breakdown:")
per_nuc_sum = 0.0
for i, nuc in enumerate(nuclide_names):
    val = float(nuc_mean[i])
    std = float(nuc_std[i])
    per_nuc_sum += val
    print(f"    {nuc:>4s}: {val:.6e} +/- {std:.6e}")

print(f"\n  Sum of per-nuclide:  {per_nuc_sum:.6e}")
print(f"  Total (from tally):  {float(total_mean[0]):.6e}")
diff = abs(per_nuc_sum - float(total_mean[0]))
print(f"  Absolute difference: {diff:.6e}")
if float(total_mean[0]) > 0:
    print(f"  Relative difference: {diff / float(total_mean[0]):.4e}")

File: crates/yamc/examples/tbr.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Instant;
use yamc::geo::{BoundaryType, Surface, SurfaceKind};
use yamc::geo::{HalfspaceType, Region, RegionExpr};
use yamc::geometry::cell::Cell;
use yamc::geometry::Geometry;
use yamc::model::{Model, TransportSettings};
use yamc_materials::Material;
use yamc_source::distribution::angular::AngularDistribution;
use yamc_source::distribution::energy::Discrete;
use yamc_source::source::{ParticleSource, Source, SourceEnergyDistribution};
use yamc_tallies::filter::Filter;
use yamc_tallies::tally::{Mt, ReactionRateScore, Score, Tally};
use yamc_tallies::CellFilter;

fn main() {
    // Create two-cell geometry: inner sphere (Li6) and outer annular region (Be9)
    let sphere1 = Arc::new(Surface {
        surface_id: Some(1),
        kind: SurfaceKind::Sphere {
            x0: 0.0,
            y0: 0.0,
            z0: 0.0,
            radius: 1.0,
        },
        boundary: BoundaryType::Transmission,
        name: None,
    });

    let sphere2 = Arc::new(Surface {
        surface_id: Some(2),
        kind: SurfaceKind::Sphere {
            x0: 0.0,
            y0: 0.0,
            z0: 0.0,
            radius: 200.0,
        },
        boundary: BoundaryType::Vacuum,
        name: None,
    });

    // Create regions
    // region1: -sphere1 (inside sphere1)
    let region1 = Region {
        expr: RegionExpr::Complement(Box::new(RegionExpr::Halfspace(HalfspaceType::Above(
            sphere1.clone(),
        )))),
    };

    // region2: +sphere1 & -sphere2 (between sphere1 and sphere2)
    let region2 = Region {
        expr: RegionExpr::Intersection(
            Box::new(RegionExpr::Halfspace(HalfspaceType::Above(sphere1.clone()))),
            Box::new(RegionExpr::Complement(Box::new(RegionExpr::Halfspace(
                HalfspaceType::Above(sphere2.clone()),
            )))),
        ),
    };

    // Create materials with different absorption characteristics
    let mut material1 = Material::new(
        HashMap::from([
            ("Li6".into(), 0.07 / 2.0), // Li4SiO4
            ("Li7".into(), 0.93 / 2.0),
            ("Be9".into(), 0.5),
        ]),
        "atom",
        "g/cm3",
        Some(2.0),
    )
    .unwrap();
    material1.set_material_id(1); // Set material_id for MaterialFilter testing
    material1.set_temperature("294");

    let mut nuclide_map = HashMap::new();
    nuclide_map.insert("Be9".to_string(), "tests/Be9.arrow".to_string());
    nuclide_map.insert("Li6".to_string(), "tests/Li6.arrow".to_string());
    nuclide_map.insert("Li7".to_string(), "tests/Li7.arrow".to_string());
    material1.read_nuclear_data(&nuclide_map, None).unwrap();

    let mat_arc = Arc::new(material1);

    // Create cells
    let cell1 = Cell::new(
        Some(1),
        region1,
        Some("inner_sphere".to_string()),
        None, // No material fill
    );

    let cell2 = Cell::new(Some(2), region2, Some("outer_annular".to_string()), Some(0));

    let geometry = Geometry::new(vec![cell1.clone(), cell2.clone()], vec![mat_arc]).unwrap();

    // Create source
    let source = ParticleSource::Neutron(Source {
        space: yamc_source::source::SourceSpatialDistribution::Point(
            yamc_source::distribution::spatial::Point::new([0.0, 0.0, 0.0]),
        ),
        angle: AngularDistribution::Isotropic,
        energy: SourceEnergyDistribution::Discrete(
            Discrete::new(vec![14.06e6], vec![1.0]).unwrap(),
        ),
        strength: 1.0,
    });

    let particles = 50000;
    let batches = 2;

    // Create tallies with CellFilters
    let cell_filter2 = Filter::Cell(CellFilter::from_id(cell2.cell_id.unwrap()));
    let mut tally1 = Tally::new();
    tally1.filters = vec![cell_filter2];
    tally1.scores = vec![Score::ReactionRate(ReactionRateScore::from_mt(Mt::new(
        105,
    )))]; // n,t (tritium production)
    tally1.name = Some("tbr".to_string());

    // Initialize batch data
    tally1.initialize_batches(batches);

    let tallies = vec![Arc::new(tally1)];

    let mut model = Model::new(geometry, vec![source], tallies);

    let start = Instant::now();
    model
        .simulate_transport(&TransportSettings {
            total_particles: Some(particles * batches),
            ..Default::default()
        })
        .unwrap();
    let elapsed = start.elapsed();

    // Check if we're the root rank (works with or without MPI)
    let is_root = yamc::mpi_context::mpi_rank() == 0;

    // Only root rank prints the results
    if is_root {
        println!(
            "Simulation completed in {:.6} seconds.",
            elapsed.as_secs_f64()
        );

        // Tallies are updated in place!
        let mean = model.tallies[0].get_mean();
        println!("TBR (tritium breeding ratio): {}", mean[0]);
    }

    // Properly finalize MPI to avoid spurious error messages
    yamc::mpi_context::mpi_finalize();
}
